### PR TITLE
refactor(coinjoin): refactor Wallet

### PR DESCRIPTION
**Pull request by HashEngineering**
## What was done?

* Refactor CoinJoin keychain code into `CoinJoinExtension`
* Add CoinJoin send function
* Fix some issues with connectivity

## Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have made corresponding changes to the documentation

**For repository code-owners and collaborators only**
- [ ] I have assigned this pull request to a milestone

### CODE REVIEW

File: core/src/main/proto/wallet.proto [hunk 1]
```diff
@@ -512,3 +512,10 @@ message ExchangeRate {
 
   // Next tag: 4
 }
+/** CoinJoin KeyChain */
+message CoinJoin {
+  // key chain
+  repeated Key key = 1;
+  // last value of rounds from mixing
+  required int32 rounds = 2;
+}
```
Comment (HashEngineering) [inline core/src/main/proto/wallet.proto:521]: `rounds` is used to calculate the balance of coinjoin coins.

if all coins were mixed for 4 rounds, but the `rounds = 5` then the coinjoin balance will be zero.  This value needs to persist.

File: core/src/main/java/org/bitcoinj/wallet/WalletEx.java [hunk 2]
```diff
@@ -0,0 +1,1150 @@
...
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import static org.bitcoinj.core.NetworkParameters.MAX_MONEY;
+
+public class WalletEx extends Wallet {
+    private static final Logger log = LoggerFactory.getLogger(WalletEx.class);
```
Comment (HashEngineering) [inline core/src/main/java/org/bitcoinj/wallet/WalletEx.java:48]: `WalletEx` is a temporary name.

File: core/src/test/java/org/bitcoinj/testing/TestWithNetworkConnections.java [hunk 3]
```diff
@@ -93,7 +93,7 @@ public void setUp(BlockStore blockStore) throws Exception {
             // Reduce the number of keys we need to work with to speed up these tests.
             KeyChainGroup kcg = KeyChainGroup.builder(UNITTEST).lookaheadSize(4).lookaheadThreshold(2).
                     fromRandom(Script.ScriptType.P2PKH).build();
-            wallet = new Wallet(UNITTEST, kcg);
+            wallet = createWallet(kcg);
```
Comment (HashEngineering) [inline core/src/test/java/org/bitcoinj/testing/TestWithNetworkConnections.java:96]: We need to allow derived classes to create their own wallet objects.

**Review by Syn-McJ — APPROVED** [review]
Looks like great work.

**Pull request merged**